Engines/Wine/Engine/Versions/script.js

```diff
@@ -1,8 +1,138 @@
+    versions.sort((a, b) =>
+    {
+        // check version format
+        const versionRegExp = /^(\d+\.\d+(\.\d+)?)(.*)?$/;
+        if (!versionRegExp.test(a.version)) {
+            throw new Error(`invalid Wine version "${a.version}`);
+        }
+        if (!versionRegExp.test(b.version)) {
+            throw new Error(`invalid Wine version "${a.version}`);
+        }
+
+        // sort
+        const aVersionParts = a.version.match(versionRegExp);
+        const bVersionParts = b.version.match(versionRegExp);
+
+        const aVersionNumbers = Array.from(aVersionParts[1].split('.')).map(item => Number(item));
+        const bVersionNumbers = Array.from(bVersionParts[1].split('.')).map(item => Number(item));
+
+        // ensure that cases where not all version numbers (major, minor, patch) are set are handled correctly
+        const maxVersionIdx = 2;
+        for (let i = 0; i <= maxVersionIdx; i++) {
+            if (typeof aVersionNumbers[i] === 'undefined') {
+                aVersionNumbers[i] = 0;
+            }
+            if (typeof bVersionNumbers[i] === 'undefined') {
+                bVersionNumbers[i] = 0;
+            }
+        }
+
+        // ensure that cases where the description is not set are handled correctly
+        let aVersionDescription = aVersionParts[3];
+        if (typeof aVersionDescription === 'undefined') {
+            aVersionDescription = "";
+        }
+
+        let bVersionDescription = bVersionParts[3];
+        if (typeof bVersionDescription === 'undefined') {
+            bVersionDescription = "";
+        }
+
+        // major
+        if (aVersionNumbers[0] != bVersionNumbers[0]) {
+            return aVersionNumbers[0] - bVersionNumbers[0];
+        }
+        // minor
+        if (aVersionNumbers[1] != bVersionNumbers[1]) {
+            return aVersionNumbers[1] - bVersionNumbers[1];
+        }
+        // patch
+        if (aVersionNumbers[2] != bVersionNumbers[2]) {
+            return aVersionNumbers[2] - bVersionNumbers[2];
+        }
+        // description
+        if (aVersionDescription < bVersionDescription) {
+            return -1;
+        } else {
+            return 1;
+        }
+    }
+    );
+}
+
+/**
+ * Fetches the latest available Wine version based on the category and version regex
+ * @param {wizard} wizard setup wizard to show the download progress of the versions json
+ * @param {string} category (e.g. "upstream-linux-x86")
+ * @param {RegExp} regex regex the version must fulfill
+ * @returns {void}
+ */
+function getLatestVersion(wizard, category, regex) {
+    const versionsJson = module.getAvailableVersions(wizard);
+
+    const packages = versionsJson
+        .filter(distribution => distribution.name === category)
+        .flatMap(distribution => distribution.packages);
+
+    const regExp = new RegExp(regex);
+    const versions = packages
+        .filter(({version}) => regExp.test(version))
+        .flatMap(package => package.version);
```

Comment:
`.flatMap(package => package.version)` or `.map(package => package.version)`?

Reply:
I'm not 100% sure (I'm not that familiar with this stuff). In any case it's working.

Reply:
`map` applies a function on a given list/array and returns a list/array with the results. E.g. if you do the following 

```javascript
[1,2,3].map(i => [i, i])
```

you will get `[[1,1], [2,2], [3,3]]`

`flatMap` also applies a function on a given list/array and returns a list/array with the results, but differing from `map` the function needs to return a list/array and the final list/array will be "flattened". This will lead to a different result for:

```javascript
[1,2,3].flatMap(i => [i, i])
```

you will get `[1,1, 2,2, 3,3]`

So the main question is: does `package.version` contain an array or a simple string?

Reply:
It's a simple string. From my understanding, `map` and `flatMap` should behave the same way in this case.

Reply:
In Java it would not work. You would get a compile time error, because the API expects the function to return a `Stream` (i.e. a collection) and not a single value.

Maybe it is a particularity of Javascript?

Reply:
I'm not sure. Just tried `map` and it works as well. I will change it.